Use Case Insensitive Filters for SLD Output

geographika opened the pull request:
See issue at #5526. 
This pull request makes checks for operators case-insensitive so "AND", "and", and "Annd" in a Mapfile expression will all produce an SLD filter. 

The functions `strcasecmp`, `strcasestr`, and `toupper` are used to ensure this. 

A new test Mapfile has been added to ensure these filters are created in each case. 
TODO - add test for brackets e.g. `AND(`

geographika commented:
There are some memory leaks showing up in Travis. See https://travis-ci.org/mapserver/mapserver/jobs/639062323?utm_source=github_status
I'm not sure if this is due to my changes, or due to previously untested codepaths now being tested. 

For example:

```
Direct leak of 48 byte(s) in 1 object(s) allocated from:

    #0 0x4bb533 in __interceptor_malloc /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:67:3
    #1 0x2b28ecb6e84f in FLTCreateFilterEncodingNode /home/travis/build/mapserver/mapserver/mapogcfilter.c:970:25
    #2 0x2b28ecd088b0 in BuildExpressionTree /home/travis/build/mapserver/mapserver/mapogcsld.c:4835:18
    #3 0x2b28ecd0afb4 in msSLDParseLogicalExpression /home/travis/build/mapserver/mapserver/mapogcsld.c:4968:12
    #4 0x2b28ecd0df4c in msSLDGetFilter /home/travis/build/mapserver/mapserver/mapogcsld.c:5166:19
    #5 0x2b28eccfd066 in msSLDGenerateUserStyle /home/travis/build/mapserver/mapserver/mapogcsld.c:4150:21
    #6 0x2b28eccefbb1 in msSLDGenerateSLDLayer /home/travis/build/mapserver/mapserver/mapogcsld.c:4314:9
    #7 0x2b28eccee7d3 in msSLDGenerateSLD /home/travis/build/mapserver/mapserver/mapogcsld.c:3196:18

```

rouault commented:
> I'm not sure if this is due to my changes, or due to previously untested codepaths now being tested.

Likely the later. I believe that adding CPLDestroyXMLNode(psNode) at https://github.com/mapserver/mapserver/blob/e04350e23843cf3c7749aeac25cc7c63fe7e1180/mapogcsld.c#L5000 should fix it

geographika commented:
@rouault - thanks for the hint. I added in https://github.com/mapserver/mapserver/pull/5980/commits/547c5b2e191f19ab11dacd362a2c5f5b01d43779 but [leaks still reported](https://travis-ci.org/mapserver/mapserver/jobs/640196399?utm_medium=notification&utm_source=github_status). 

Also failing to [build in Appveyor](https://ci.appveyor.com/project/MapServer/mapserver/builds/30270640), is there some casting that needs to be done?

```
C:\projects\mapserver\mapogcsld.c(4999): error C2220: warning treated as error - no 'object' file generated [C:\projects\mapserver\build\mapserver.vcxproj]
C:\projects\mapserver\mapogcsld.c(4999): warning C4133: 'function': incompatible types - from 'FilterEncodingNode *' to 'CPLXMLNode *' [C:\projects\mapserver\build\mapserver.vcxproj]
```
I removed the erroneous `free(pszTmp);` I added as per your comment.

rouault commented:
>  is there some casting that needs to be done?

ah my mistake...  `FLTFreeFilterEncodingNode(psNode);` should do it hopefully

geographika commented:
@rouault - thanks for that - updated. I also needed to free the left and right expression variables as at https://github.com/mapserver/mapserver/pull/5980/commits/79f66b6b8f18ce08ca99fb63ca18f0850dd3f7e7 and now all reported leaks are gone. 

The existing parsing used in the code looks like it could have some edge cases that would cause it to fail e.g. if an expression contained `NOR(` it would be picked up as an `OR` (although this situation may not be possible). There is also an issue (present in the existing code) where attribute values aren't trimmed - I can open a separate issue about this. 

This pull request fixes issues around case-sensitivity (previously no filters were created for these), adds some tests for the SLD creation functions, fixes some comment typos, and fixes a couple of memory leaks.

rouault commented on mapogcsld.c:
```diff
@@ -4993,8 +4996,8 @@ char *msSLDParseLogicalExpression(char *pszExpression, const char *pszWfsFilter)
 
       free(pszFLTExpression);
       pszFLTExpression = pszTmp;
+      FLTFreeFilterEncodingNode(psNode);
```
this should be moved just after the } of line 5000, so that psNode is also freed when != NULL but when pszFLTExpression is NULL

geographika commented:
@rouault - many thanks for the review and explanations
